### PR TITLE
[SPARK-29727][ML][PYTHON] Use UnaryTransformer as parent class for single input column transformers in PySpark

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Use UnaryTransformer as parent class for single input column transformers in PySpark

### Why are the changes needed?
In Scala, UnaryTransformer is the parent class of all the single input column transformers. In PySpark, we have the corresponding Python version of UnaryTransformer but it's not really used. This Jira will make single input column transformers in Python to use UnaryTransformer as parent class. 

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Use existing tests.

### CODE REVIEW

File: python/pyspark/ml/feature.py [hunk 1]
```diff
@@ -1014,10 +1015,21 @@ def setOutputCol(self, value):
...
+        return self._java_obj.createTransformFunc()
+
+    def outputDataType(self):
+        self._transfer_params_to_java()
+        return self._java_obj.outputDataType()
+
+    def validateInputType(self, inputType):
+        self._transfer_params_to_java()
+        self._java_obj.validateInputType(inputType)
+
```
Comment (huaxingao) [inline python/pyspark/ml/feature.py]: Do we still need to support Python 2.7? I have to implement the above three methods to make Python 2.7 pass. I tried to put _ in front of these method names, but 2.7 failed with ```TypeError: Can't instantiate abstract class DCT with abstract methods createTransformFunc, outputDataType, validateInputType```

**Comment by huaxingao** [comment]
@srowen @zhengruifeng 
Could you please review? Thanks!

File: python/pyspark/ml/feature.py (hunk 1 above)
Comment (srowen) [reply python/pyspark/ml/feature.py]: For now, yeah we need to support Python 2, still

**Review by srowen — COMMENTED** [review]
Again this doesn't change behavior, just refactors?
You both know more about this than I do, but if that's the case, it seems OK to me.

**Comment by huaxingao** [comment]
@srowen Thanks for your review. Yes, this doesn't change behavior, just refactors. 
I looked the code again, I think actually I need to add _single_leading_underscore in front of the method names to indicate internal usage. Will add now.

**Comment by BryanCutler** [comment]
@huaxingao I'm not sure it makes sense to add a `JavaUnaryTransformer`. The additional things it does, like check the schema and add the transformed column are already done in the scala version of these transformers right?

**Comment by huaxingao** [comment]
@BryanCutler 
Thanks for your comment. Seems if I want to use UnaryTransformer as the parent class for DCT etc. , I will need to have a ```JavaUnaryTransformer```, right? I can't use ```UnaryTransformer``` directly like ```class DCT(UnaryTransformer, JavaMLReadable, JavaMLWritable)``` because I need the ```_java_obj```.

**Comment by huaxingao** [comment]
I will change the three methods to no-op as following in all the subclasses because it's no need to implement them in python side, but I have to implement to make 2.7 work, otherwise, 2.7 throws Exception ```TypeError: Can't instantiate abstract class DCT with abstract methods createTransformFunc, outputDataType, validateInputType```
```
    def _createTransformFunc(self):
        pass

    def _outputDataType(self):
        pass

    def _validateInputType(self, inputType):
        pass
```

**Comment by BryanCutler** [comment]
>Thanks for your comment. Seems if I want to use UnaryTransformer as the parent class for DCT etc. , I will need to have a JavaUnaryTransformer, right? 

My point is why do you want to use the pyspark UnaryTransformer? Before, it would 'transformShema` and call the `udf` from Scala, which is ideal. With this change the pyspark `UnaryTransformer` does those in python and uses a python udf right?

**Comment by huaxingao** [comment]
@BryanCutler You are right. I will close this PR. Thanks for your review!

**Comment by BryanCutler** [comment]
Ok thanks @huaxingao . Are there any tests using pyspark `UnaryTransformer`?  If not it might be good to add one to show proper usage.

**Comment by huaxingao** [comment]
@BryanCutler Yes. In PySpark tests, there is a ```MockUnaryTransformer``` to show the usage of ```UnaryTransformer```. One test to cover the methods in ```MockUnaryTransformer``` and another one to test the persistence.  These should be good enough.

**Pull request closed without merging**